### ISSUE TITLE
Performance Data of Hexo

### ISSUE THREAD

**Issue by SukkaW** [hexojs/hexo#3663]
# Check List

Please check followings before submitting a new issue.

- [x] I have already confirmed other issue template and they are not different my want

## Environment & Settings

**Node.js & npm version**

Test with 8.16.0, 10.16.2 and 12.8.0

**Your package.json `package.json`**

https://github.com/theme-suka/hexo-theme-unit-test/blob/master/package.json

## Others

I have ran a performance test for Hexo using Travis CI with [300 posts](https://github.com/SukkaLab/hexo-many-posts)(All posts have 1 category and 5 tags, and all markdown styles and large amounts of code blocks are included in each .md file)

All the css, js & images are deleted before `hexo clean && hexo g` to make sure every file generated is rendered HTML.

| | NodeJS 8 | NodeJS 10 | NodeJS 12 |
|---|:---:|:---:|:---:|
| Enable Fragment Cache & Disable Highlight.js | 26.5s | 21s | 21.5s |
| Disable Fragment Cache | 35s | 29s | 29.5s |
| Enable highlight.js only | 35s | 27s | 27.5s |
| Enable highlight.js & line_number | 38s | 30s | 30s |
| Enable highlight.js & tab_replace | 34.5s | 27s | 28s |
| Enable highlight.js, tab_replace & line_number | 38s | 30s | 31s |

Raw log can be seen [here](https://travis-ci.com/theme-suka/performance-test).

It seems that `Disable Fragment Cache` will make hexo `35%` slower, `Enable highlight.js` will make hexo `45%` slower.
And even with Fragment enabled and Highlight.js disabled, hexo can only render 33 HTML per second.

**Comment by tcrowe** [comment]
I had not heard of that setting so I wanted to help provide some context:
https://hexo.io/docs/helpers#fragment-cache
https://hexo.io/docs/templates#Optimization
https://github.com/search?q=org%3Ahexojs+fragment&type=Code

**Comment by SukkaW** [comment]
@tcatche Fragment Cache was introduced in https://github.com/hexojs/hexo/pull/637 and the feature is depends on theme.
And Fragment Cache will be much more helpful if some time-consuming partial has been cached, seen https://github.com/hexojs/hexo/issues/1769

**Comment by ppoffice** [comment]
@SukkaW Please comment out this line, rerun your tests and see how results are

https://github.com/hexojs/hexo/blob/master/lib/plugins/filter/index.js#L14

**Comment by SukkaW** [comment]
@ppoffice Yeah, I do have received the notice that `meta_generator` has contributed to poor performance. I will add a test with `meta_generator` disabled and see what we can get.

**Comment by SukkaW** [comment]
@ppoffice The result is out:

![image](https://user-images.githubusercontent.com/40715044/63068325-5454c100-bf45-11e9-965b-9f7a63337fb4.png)

This is baseline

![image](https://user-images.githubusercontent.com/40715044/63068333-5e76bf80-bf45-11e9-9762-86eb71b8663a.png)

Set `meta_generator: false` in `_config.yml`

![image](https://user-images.githubusercontent.com/40715044/63068367-82d29c00-bf45-11e9-891e-0738c44ba501.png)

Comment out the `filter.register('after_render:html', require('./meta_generator'));`

**Comment by SukkaW** [comment]
After having a look at git blame, I found out that the `meta_generator` was introduced in https://github.com/hexojs/hexo/pull/3129.

**Comment by ppoffice** [comment]
@SukkaW `meta_generator` config was introduced 9 days ago. You probably need to update your Hexo in your test to make the `meta_generator: false` work.

https://github.com/hexojs/hexo/pull/3653

**Comment by SukkaW** [comment]
@ppoffice 
https://github.com/hexojs/hexo/blob/master/lib/plugins/filter/after_post_render/external_link.js#L8
Or like this, return before `cheerio.load`

**Comment by jiangtj** [comment]
https://gitlab.com/JiangTJ/performance-test-hexo/-/jobs/272854627

When there are too many tags, some of the theme code can lead to very long generation times, which I think should be avoided in hexo.

See:
- https://github.com/hexojs/hexo/pull/3624
- https://github.com/theme-next/hexo-theme-next/issues/539
- https://github.com/ppoffice/hexo-theme-icarus/issues/513

**Comment by SukkaW** [comment]
> https://github.com/hexojs/hexo/blob/master/lib/plugins/filter/after_post_render/external_link.js#L8
> Or like this, return before `cheerio.load`

I have created a PR about this.

Anyway, I totally disagree to use cheerio traversing all the HTML generated, just for add an `meta_generator` tag. There should be some better ways to do that.

**Comment by curbengh** [comment]
> Anyway, I totally disagree to use cheerio traversing all the HTML generated, just for add an `meta_generator` tag. There should be some better ways to do that.

[open_graph](https://github.com/hexojs/hexo/blob/master/lib/plugins/helper/open_graph.js) approach?

This also means (with that approach) `meta_generator` is no longer widely adopted (across Hexo), as it depends how soon a theme can adopt it and how often users update their theme. I believe themes will adopt it eventually, albeit will take some time to reach users.

Ultimately, it depends how much we care about [Wappalyzer statistics](https://www.wappalyzer.com/categories/static-site-generator). Judging by the performance benchmark, maybe it's not worth it.

---

I'm very interested if commenting out the `<%- open_graph() %>` from theme can improve performance.

**Comment by SukkaW** [comment]
@curbengh Currently the theme I used in performance test, my [hexo-theme-suka](https://github.com/SukkaW/hexo-theme-suka/blob/master/layout/_partial/head/open-graph.ejs), Open Graph is generated manually without using Hexo helper.

Maybe I would bring up a test with hexo-theme-landscape to test the performance of open-graph helper.

Also, cheerio is very slow... Especially when it is needed to let cheerio ran through all the hundreds of, maybe thousands of  HTML files generated.
We should came up with a better idea to detect and add `meta[generator]` tag.

**Comment by SukkaW** [comment]
@curbengh 

I just bring up the benchmark for `open_graph()` helper.

Raw test log can be found here: https://travis-ci.com/theme-suka/performance-test/builds/123338750

**Teat A**

with `hexo-theme-landscape` and `meta_generator` set to false.

![image](https://user-images.githubusercontent.com/40715044/63083284-e88d4b00-bf7a-11e9-8e85-ebf2f04eff64.png)

**Teat B**

with `hexo-theme-landscape` and `meta_generator` set to false, and remove `open_graph()` helper using:

```bash
sed -i "s|<%- open_graph({twitter_id: theme.twitter, fb_admins: theme.fb_admins, fb_app_id: theme.fb_app_id}) %>||g" themes/landscape/layout/_partial/head.ejs
```

![image](https://user-images.githubusercontent.com/40715044/63083361-1f636100-bf7b-11e9-8949-ad6042f176c5.png)

It is 28% faster without `open_graph()` helper.

**Comment by curbengh** [comment]
I wonder if cheerio@1.0.0rc.3 is faster. It's [currently unusable](https://github.com/cheeriojs/cheerio/issues/1198) for non-ASCII chars though, but it shouldn't affect the benchmarking if we disregard how the rendered file looks like.

If you need the rendered file to be valid, you can use `"cheerio": "curbengh/cheerio#decode-test"` which includes [the fix](https://github.com/cheeriojs/dom-serializer/pull/85). you may also need to enable [`decodeEntities`](https://github.com/curbengh/hexo-nofollow/blob/b720cd7ec1e05a9a36a3d265c13eec39e4e8f2c6/lib/filter.js#L42-L44).

**Comment by SukkaW** [comment]
@curbengh 
For `open_graph()` helper I have an idea. Since cheerio is only used to list all the images in the post,  we might use `indexOf('<img')` first to determine if the current content has image or not.
If the current post has no images, then cheerio is no need to load the whole post.

**Comment by curbengh** [comment]
> @curbengh
> For `open_graph()` helper I have an idea. Since cheerio is only used to list all the images in the post at `open_graph()`, we could use `indexOf('<img')` first to determine if the current content has image or not.
> If the current post has no images, then cheerio is no need to load the whole post.

I like the idea, though ES6 `includes(substring)` is preferred. Looking forward for your PR.

> which means even without images in the content, cheerio might still slow down the speed.

your proposed fix is supposed to fix it, yes?

**Comment by SukkaW** [comment]
A benchmark of `load cheerio when needed`:

**Teat A**

with hexo-theme-landscape and meta_generator set to false, with [cheer io is always required](https://github.com/hexojs/hexo/blob/master/lib/plugins/helper/open_graph.js#L64).

![image](https://user-images.githubusercontent.com/40715044/63095345-9dd0fa80-bf9d-11e9-8743-9a85ae5297d2.png)

**Teat B**

with hexo-theme-landscape and meta_generator set to false, and only require cheerio when `content` includes `<img`.

```javascript
  if (!images.length && content) {
    images = images.slice();

    if (content.includes('<img')) {
      if (!cheerio) cheerio = require('cheerio');

      const $ = cheerio.load(content);

      $('img').each(function() {
        const src = $(this).attr('src');
        if (src) images.push(src);
      });
    }
  }
```

![image](https://user-images.githubusercontent.com/40715044/63095451-da045b00-bf9d-11e9-8402-7ab3578e6469.png)

As those 300 posts includes no images, generate speed is now `17%` faster. I will open a PR then.

**Comment by tcrowe** [comment]
@ppoffice Is it possible you can show me an example of this? `how Hexo generator spawn promises`

Also we hardly need cheerio for simple dom operations. Possibly we can just use string templates for the simple cases.

**Comment by ppoffice** [comment]
@tcrowe #3665

**Comment by SukkaW** [comment]
@tcrowe I have listed the helpers & filters that use cheerio. Maybe we can replace them with regex.
https://github.com/hexojs/hexo/issues/3663#issuecomment-521489430

**Comment by SukkaW** [comment]
I'm also interested in the performance impact of `external_link` filter, so I bring up a benchmark.

**Test A**

Baseline, `external_link` is set to true.

![image](https://user-images.githubusercontent.com/40715044/63138972-0ad0a880-c00f-11e9-8ae3-04a6d6d00e94.png)

**Test B**

`external_link` is set to false.

![image](https://user-images.githubusercontent.com/40715044/63139024-4ec3ad80-c00f-11e9-9240-e4b6258a2210.png)

I can see only a little improvements on generating speed, less than `0.5%`.

It might because `external_link` filter is registered at `after_post_render`, which means cheerio only needs to load post content. And as `meta_generator` filter registered at `after_render:html`, cheerio needs to load all the HTML files generated (usually double to three times the number of posts due to tags and categories), that's why it caused a serious performance impact.

**Comment by curbengh** [comment]
> It might because `external_link` filter is registered at `after_post_render`, which means cheerio only needs to load post content. And as `meta_generator` filter registered at `after_render:html`, cheerio needs to load all the HTML files generated (usually double to three times the number of posts due to tags and categories), that's why it caused a serious performance impact.

I just found `after_post_render` only works on post, not pages. So, it is appropriate for meta_generator to be `after_render:html`. `external_link` filter should also be `after_render:html` as well.

**Comment by SukkaW** [comment]
I have noticed that during those benchmarks, my `hexo-theme-suka` is always `10%` slower than `hexo-theme-landscape`. And I have found that `hexo-theme-suka` uses `toc()` helper and `hexo-theme-landscape` doesn't. So I bring up a new test with `hexo-theme-suka`:

**Test A**

Baseline, with `meta_generator` commented out.

![image](https://user-images.githubusercontent.com/40715044/63222144-16220080-c1d6-11e9-8856-7dc4938fca5a.png)

**Test B**

Disable the toc.

![image](https://user-images.githubusercontent.com/40715044/63222145-1ae6b480-c1d6-11e9-8d47-28f4635d903d.png)

Bingo! `toc()` helper slowed down the generation by `24.6%`.

Then I looked at the `toc()` helper and have some interesting discovery.

https://github.com/hexojs/hexo/blob/49c7d4a897f15cdeffeccf9d5d7772ae616f2969/lib/plugins/helper/toc.js#L9-L14

Most of the theme use `toc()` helper like this:

```javascript
toc(page.content)
```

which means when a theme is using `toc()` helper, the cheerio will load all posts' content. In the benchmarks, `toc()` helper will load all 300 posts content.
